src/thermo/Phase.cpp

```diff
@@ -643,6 +761,22 @@ doublereal Phase::molarVolume() const
     return 1.0/molarDensity();
 }
 
+void Phase::setDensity(const doublereal density_)
+{
+    assertCompressible("setDensity");
+    if (density_ > 0.0) {
+        m_dens = density_;
+    } else {
+        throw CanteraError("Phase::setDensity()",
+            "density must be positive. density = {}", density_);
+    }
+}
+
+void Phase::setConstDensity(const doublereal density_)
+{
+    m_dens = density_;
+}
```

Comment:
I think this should at least do the same sanity check as `setDensity` (i.e. that the density is positive). It might also make sense to require that the phase is actually incompressible.

I'm wondering if there's a way we can make this function `protected`, given that almost all of its use is internal to Phase and derived classes -- the only exception is the `LatticeSolidPhase.fromScratch` test, which is a bit arbitrary anyway.

Reply:
NP on the check. For `LatticeSolidPhase.fromScratch` I have no opinion. (I'm ok with it being `pretected`). 

On a similar note, I made `setStoichMoleFractions` protected. Not sure what happened to your original comment, but my response was that it was needed to avoid a warning raised by the regular pathway.

Reply:
On a related note: other phases (`StoichSubstance` etc.) side-step this issue by using `setParameters`